sdk/webpubsub/Microsoft.Azure.WebJobs.Extensions.WebPubSub/src/Utilities.cs

```diff
@@ -77,80 +76,89 @@ public static HttpResponseMessage BuildConnectEventResponse(string response, Dic
-                originStr = response.ToString();
-                converted = JsonDocument.Parse(originStr);
+                if (response is UserEventResponse messageResponse)
+                {
+                    var mergedStates = context.UpdateStates(messageResponse.States);
+                    return BuildUserEventResponse(messageResponse, mergedStates);
+                }
+                return BuildErrorResponse($"Invalid response type: '{response.GetType()}' in current request type '{requestType}'.");
             }
+            // should not hit.
+            throw new Exception($"Invalid request type, {requestType}");
+        }
 
+        public static HttpResponseMessage BuildValidResponse(
+            JToken jResponse, RequestType requestType,
+            WebPubSubConnectionContext context)
+        {
             try
             {
-                // Check error, errorCode is required for json convert, otherwise, ignored.
-                if (needConvert && converted.RootElement.TryGetProperty("code", out var code))
+                // Accepts JObject response or stringified, otherwise return ServerError.
+                JObject response = jResponse is JObject jObj ? jObj : JObject.Parse(jResponse.ToString());
+
+                // check error as top priority.
+                if (response.TryGetValue("code", out var code)
+                    && code.ToObject<WebPubSubStatusCode>() != WebPubSubStatusCode.Success)
                 {
-                    var error = JsonSerializer.Deserialize<EventErrorResponse>(originStr);
+                    var error = response.ToObject<EventErrorResponse>();
                     return BuildErrorResponse(error);
                 }
-                else if (response is EventErrorResponse errorResponse)
-                {
-                    return BuildErrorResponse(errorResponse);
-                }
 
                 if (requestType == RequestType.Connect)
                 {
-                    if (needConvert)
-                    {
-                        var states = GetStatesFromJson(converted, originStr);
-                        var mergedStates = context.UpdateStates(states);
-                        return BuildConnectEventResponse(originStr, mergedStates);
-                    }
-                    else if (response is ConnectEventResponse connectResponse)
-                    {
-                        var mergedStates = context.UpdateStates(connectResponse.States);
-                        return BuildConnectEventResponse(JsonSerializer.Serialize(response), mergedStates);
-                    }
+                    var states = GetStatesFromJson(response);
+                    var mergedStates = context.UpdateStates(states);
+                    // Use original value directly to set response.
+                    return BuildConnectEventResponse(JsonConvert.SerializeObject(response), mergedStates);
```

Comment:
is it safer to convert the response to `ConnectEventResponse` first? Since the raw jobject `response` might contain additional fields that we don't want to include in the response, for example, when customer returns:
```js
{
   "dummy": "abc"
}
```